server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java

```diff
@@ -0,0 +1,336 @@
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.query;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.CollectionTerminatedException;
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.FilterScorable;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.Scorable;
+import org.apache.lucene.search.ScoreCachingWrappingScorer;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.ScorerSupplier;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
+import org.elasticsearch.common.lucene.Lucene;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Top-level collector used in the query phase to perform top hits collection as well as aggs collection.
+ * Inspired by {@link org.apache.lucene.search.MultiCollector} but specialized for wrapping two collectors and filtering collected
+ * documents as follows:
+ * - through an optional <code>post_filter</code> that is applied to the top hits collection
+ * - through an optional <code>min_score</code> threshold, which is applied to both the top hits as well as aggs.
+ * Supports also terminating the collection after a certain number of documents have been collected (<code>terminate_after</code>).
+ *
+ * When top docs as well as aggs are collected (because both collectors were provided), skipping low scoring hits via
+ * {@link Scorable#setMinCompetitiveScore(float)} is not supported for either of the collectors.
+ */
+final class QueryPhaseCollector implements Collector {
+    private final Collector aggsCollector;
+    private final Collector topDocsCollector;
+    private final int terminateAfter;
+    private final Weight postFilterWeight;
+    private final Float minScore;
+    private final boolean cacheScores;
+
+    private int numCollected;
+    private boolean terminatedAfter = false;
+
+    QueryPhaseCollector(Collector topDocsCollector, Weight postFilterWeight, int terminateAfter, Collector aggsCollector, Float minScore) {
+        this.topDocsCollector = Objects.requireNonNull(topDocsCollector);
+        this.postFilterWeight = postFilterWeight;
+        if (terminateAfter < 0) {
+            throw new IllegalArgumentException("terminateAfter must be greater than 0");
+        }
+        this.terminateAfter = terminateAfter;
+        this.aggsCollector = aggsCollector;
+        this.minScore = minScore;
+        this.cacheScores = aggsCollector != null && topDocsCollector.scoreMode().needsScores() && aggsCollector.scoreMode().needsScores();
+    }
+
+    @Override
+    public void setWeight(Weight weight) {
+        if (postFilterWeight == null && minScore == null) {
+            // propagate the weight when we do no additional filtering over the docs that are collected
+            topDocsCollector.setWeight(weight);
+        }
```

Comment:
Maybe add a bit more context here about how collect will be called for topDocsCollector when postFilterWeight is set?